src/compression/b44/mod.rs

```diff
@@ -0,0 +1,924 @@
+    type_: SampleType,
+    quantize_linearly: bool,
+    samples_per_pixel: usize,
+}
+
+fn cpy_u16(src: &[u16], src_i: usize, dst: &mut [u16], dst_i: usize, n: usize) {
+    // assert_eq!(src.len(), dst.len());
+
+    // for i in 0..src.len() {
+    for i in 0..n {
+        dst[dst_i + i] = src[src_i + i];
+    }
+}
+
+pub fn decompress(
+    channels: &ChannelList,
+    compressed: &ByteVec,
+    rectangle: IntegerBounds,
+    expected_byte_size: usize,
+    _pedantic: bool,
+) -> Result<ByteVec> {
+    let expected_value_count = expected_byte_size / 2;
+    debug_assert_eq!(
+        expected_byte_size,
+        rectangle.size.area() * channels.bytes_per_pixel
+    );
+    debug_assert_ne!(expected_value_count, 0);
+    debug_assert!(!channels.list.is_empty());
+
+    if compressed.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let mut channel_data: Vec<ChannelData> = Vec::with_capacity(channels.list.len());
+    let mut tmp_read_index = 0;
+
+    for channel in channels.list.iter() {
+        let channel = ChannelData {
+            tmp_start_index: tmp_read_index,
+            tmp_end_index: tmp_read_index,
+            resolution: channel.subsampled_resolution(rectangle.size),
+            y_sampling: channel.sampling.y(),
+            type_: channel.sample_type,
+            quantize_linearly: channel.quantize_linearly,
+            samples_per_pixel: channel.sample_type.bytes_per_sample()
+                / SampleType::F16.bytes_per_sample(),
+        };
+
+        tmp_read_index += channel.resolution.area() * channel.samples_per_pixel;
+        channel_data.push(channel);
+    }
+
+    let mut in_i = 0usize;
+    let mut remaining = compressed.len();
+
+    let mut out = vec![0u8; expected_byte_size];
+    let mut tmp_u16_buffer = Vec::<u16>::with_capacity(expected_byte_size / 2);
+    let mut out_i = 0;
+
+    debug_assert_eq!(tmp_read_index, expected_value_count);
+
+    for channel in &mut channel_data {
+        // Sample types that does not support B44 compression (u32 and f32) are raw copied.
+        if channel.type_ != SampleType::F16 {
+            // Compute byte count for current channel.
+            let byte_count = channel.resolution.area()
+                * channel.samples_per_pixel
+                * channel.type_.bytes_per_sample();
+
+            if remaining < byte_count {
+                // notEnoughData();
+                return Err(Error::invalid("not enough data"));
+            }
+
+            //memcpy (cd.start, inPtr, n);
+            for i in channel.tmp_start_index..(channel.tmp_start_index + byte_count) {
+                // TOTALLY WRONG
+                out[out_i + i] = compressed[in_i + i];
+            }
+
+            // Raw copy bytes in u16 temp buffer.
+            for i in (0..byte_count).step_by(2) {
+                let v0 = compressed[in_i + i];
+                let v1 = compressed[in_i + i + 1];
+                tmp_u16_buffer.push(u16::from_be_bytes([v0, v1]))
```

Comment:
Challenge: Try to copy bytes into the u16 buffer without `for` and without any indexing integer. Here a some hints, start from top to bottom if you are stuck:

- [hint 1](https://doc.rust-lang.org/std/iter/trait.Extend.html)
- [hint 2](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take)
- [hint 3](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact)

Reply:
I end up with this working:

```rust
tmp_u16_buffer.extend(compressed.chunks_exact(2).map(|c| u16::from_be_bytes([c[0], c[1]])));
```

Reply:
Awesome! That's exactly what I had in mind. Of course, whether this is better than the original is still up to debate. It's a rather strange piece of algorithm, but if it works, it aint stupid.

Reply:
This code looks like it might be a single call to one of the `lebe` functions. `lebe` also supports the simple conversion from u16 to u8 bytes, even if no endianness conversion is desired. 

Remember that a vec can also be used like any `io::Write` object, which appends the written bytes to the end of the vector

Reply:
> which appends the written bytes to the end of the vector

This is my current limitation, as B44 manipulate blocks of 4x4. When I decode them, I need the final image buffer to write each of the 4 rows.

Reply:
I don't know how the data is structured. I assume all four rows are stored one after another in a large buffer? Maybe it would make sense to split the buffer into the four rows, using either `buffer.chunks_mut(row_length)` or calling `buffer.split_at_mut(nect_row_end)`. This way you can obtain four slices which can be used to write the four rows simultaneously.

Reply:
How about `tmp_bytes[..].read_from_big_endian_into(&mut u16_buffer[..])`?